core/recorder.h

```diff
@@ -0,0 +1,127 @@
+
+struct Recorder {
+    virtual ~Recorder()
+    {
+    }
+
+    /** Record an expression value as a row.  This will be flattened by
+        datasets that require flattening.
+    */
+    virtual void
+    recordRowExpr(const RowName & rowName,
+                  const ExpressionValue & expr) = 0;
+
+    /** See recordRowExpr().  This has the same effect, but takes an rvalue
+        which is destroyed by the call.  This may result in performance
+        improvements.
+    */
+    virtual void
+    recordRowExprDestructive(RowName rowName,
+                             ExpressionValue expr);
+
+    /** Record a pre-flattened row. */
+    virtual void
+    recordRow(const RowName & rowName,
+              const std::vector<std::tuple<ColumnName, CellValue, Date> > & vals) = 0;
+
+    /** See recordRow().  This has the same effect, but takes an rvalue
+        which is destroyed by the call.  This may result in performance
+        improvements.
+    */
+    virtual void
+    recordRowDestructive(RowName rowName,
+                         std::vector<std::tuple<ColumnName, CellValue, Date> > vals);
+
+    /** Record multiple flattened rows in a single transaction.  Default
+        implementation forwards to recordRow.
+    */
+    virtual void
+    recordRows(const std::vector<std::pair<RowName, std::vector<std::tuple<ColumnName, CellValue, Date> > > > & rows) = 0;
+
+    /** See recordRows().  This has the same effect, but takes an rvalue
+        which is destroyed by the call.  This may result in performance
+        improvements.
+    */
+    virtual void
+    recordRowsDestructive(std::vector<std::pair<RowName, std::vector<std::tuple<ColumnName, CellValue, Date> > > > rows);
+
+    /** Record multiple rows from ExpressionValues.  */
+    virtual void
+    recordRowsExpr(const std::vector<std::pair<RowName, ExpressionValue > > & rows) = 0;
+
+    /** See recordRowsExpr().  This has the same effect, but takes an rvalue
+        which is destroyed by the call.  This may result in performance
+        improvements.
+    */
+    virtual
+    void recordRowsExprDestructive(std::vector<std::pair<RowName, ExpressionValue > > rows);
+
+    /** Return a function specialized to record the same set of atomic values
+        over and over again into this chunk.
+
+        The returned function can be called row by row, without needing to
+        mention column names (inferred by the positions).  The values will
+        be destroyed on insertion.  numVals must always equal the number of
+        columns passed.
+
+        Extra allows for unexpected columns to be recorded, for when the
+        dataset contains some fixed columns but also may have dynamic
+        columns.
+    */
+    virtual
+    std::function<void (RowName rowName,
+                        Date timestamp,
+                        CellValue * vals,
+                        size_t numVals,
+                        std::vector<std::pair<ColumnName, CellValue> > extra)>
```

Comment:
Yes, since we will destroy it we need to pass by copy.  The caller needs to std::move the argument if it's destroyable or pass it directly (which will invoke the copy constructor) if not.  Note that the values and rowname will also be destroyed, so it's consistent over the interface.